### PR TITLE
add team selection

### PR DESCRIPTION
**Pull request by jan-Kulpas**

### CODE REVIEW

File: lib/services/team_repository.dart [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+import 'package:regatta_buddy/models/event.dart';
+import 'package:regatta_buddy/models/team.dart';
+import 'package:regatta_buddy/providers/event_details/teams_provider.dart';
+import 'package:regatta_buddy/providers/firebase_providers.dart';
+import 'package:regatta_buddy/providers/repository_providers.dart';
+
+class TeamRepository {
+  final Ref _ref;
+  final Event event;
+  //final Logger _logger = getLogger((TeamRepository).toString());
```
Comment (bartoz1) [inline lib/services/team_repository.dart:14]: logging is very important

File: lib/services/team_repository.dart [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+    final query = await _firestore.collection('events/${event.id}/teams').get();
+    return query.docs.map((e) => Team.fromDocument(e)).toList();
+  }
+
+  /// Creates a team with a set name with creating user as team captain.
+  /// Function fails is user is already in a team or is the event host.
+  /// Note that we do not sanitise the team name in any way.
+  Future<Either<String, Unit>> createTeamFromName(String name) async {
+    final uid = _ref.read(firebaseAuthProvider).currentUser?.uid;
+
```
Comment (bartoz1) [inline lib/services/team_repository.dart:48]: i would add log here, that we are creating a team

File: lib/services/team_repository.dart [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+  void updateTeam(String teamId, Team team) {
+    _firestore.collection('events/${event.id}/teams').doc(teamId).update(team.toJson());
+  }
+
+  void deleteTeam(String teamId) {
+    _firestore.collection('events/${event.id}/teams').doc(teamId).delete();
+  }
+
+  Future<Team> getTeam(String id) async {
+    final doc = await _firestore.collection('events/${event.id}/teams').doc(id).get();
```
Comment (MarekBrandt) [inline lib/services/team_repository.dart:34]: you could use getTeams() inside this one

File: lib/services/team_repository.dart [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+    return await _firestore.runTransaction((transaction) async {
+      if (await _ref.read(userRepositoryProvider).isUserInEvent(userId: uid, eventId: event.id)) {
+        return left("User is already in a team.");
+      }
+
+      // update the team
+
+      // This step is a bit tricky because it implies any user has security access to the team.
+      // Not really sure how to avoid it locally BUT we could use cloud function to update the team
+      // on the backend maybe? (also we get to write about cloud functions as inżynierka padding)
```
Comment (bartoz1) [inline lib/services/team_repository.dart:96]: i like it, we could do that in the future

File: lib/services/team_repository.dart [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+      final team = await getTeam(teamId);
+      updateTeam(teamId, team.copyWith(members: [...team.members, uid]));
+
+      // update the user joined events list
```
Comment (bartoz1) [inline lib/services/team_repository.dart:100]: could be removed

File: lib/services/user_repository.dart [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+  Future<Either<String, UserData>> getCurrentUserData() async {
+    final AuthState authState = _ref.watch(authStateNotiferProvider);
+
+    return authState.maybeWhen(
+      orElse: () => throw Exception("User not logged in"),
+      authenticated: (user) => getUserData(user.uid),
+    );
+  }
+
+  /// Checks if user is already participating in the event.
```
Comment (MarekBrandt) [inline lib/services/user_repository.dart:54]: remove this comment. Name of method is descriptive enought

File: lib/services/user_repository.dart [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+    // when accessing the top level user data.
+    //
+    // However we could reconsider using an entire collection and instead
+    // maybe have a collection called private with a single doc containing
+    // the event list?
+
+    _firestore.collection('users/$userId/joinedEvents').doc(eventId).set({'teamId': teamId});
+  }
+
+  /// Removes an event from the list of joined events
```
Comment (MarekBrandt) [inline lib/services/user_repository.dart:79]: same here

**Review by bartoz1 — APPROVED** [review]

File: lib/services/team_repository.dart (hunk 5 above)
Comment (jan-Kulpas) [reply lib/services/team_repository.dart:100]: ??

File: lib/widgets/event_teams_display.dart [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+          ),
+          loading: () => const CircularProgressIndicator(),
+          error: (error, stackTrace) => Text(error.toString()),
+        ),
+      ),
+    );
+  }
+}
+
+class CreateTeamButton extends HookConsumerWidget {
```
Comment (MarekBrandt) [inline lib/widgets/event_teams_display.dart:62]: extract other classes to their own files please

File: pubspec.yaml [hunk 9]
```diff
@@ -11,17 +11,18 @@ dependencies:
...
   cupertino_icons: ^1.0.2
-  dartz: ^0.10.1
+  fpdart: ^1.1.0
   fake_cloud_firestore: ^2.4.1+1
   firebase_auth: ^4.9.0
   firebase_core: ^2.15.1
   firebase_core_platform_interface: ^4.8.0
   firebase_database: ^10.2.5
   flutter_compass: ^0.7.0
+  flutter_hooks: ^0.20.1
```
Comment (MarekBrandt) [inline pubspec.yaml:21]: flutter hooks dodales?

**Review by MarekBrandt — APPROVED** [review]

**Pull request merged**